docs/guides/haiku_migration_guide.rst

```diff
@@ -0,0 +1,192 @@
+      self.dmid = dmid
+      self.dout = dout
+
+    def __call__(self, x, training: bool):
+      x = Block(self.dmid)(x, training)
+      x = hk.Linear(self.dout)(x)
+      return x
+
+  ---
+
+  import flax.linen as nn
+
+  class Block(nn.Module):
+    features: int
+
+
+    @nn.compact
+    def __call__(self, x, training: bool):
+      x = nn.Dense(self.features)(x)
+      x = nn.BatchNorm(
+        use_scale=True, use_bias=True, momentum=0.99
+      )(x, use_running_average=not training)
+      x = nn.Dropout(0.5, deterministic=not training)(x)
+      x = jax.nn.relu(x)
+      return x
+
+  class Model(nn.Module):
+    dmid: int
+    dout: int
+
+
+    @nn.compact
+    def __call__(self, x, training: bool):
+      x = Block(self.dmid)(x, training)
+      x = nn.Dense(self.dout)(x)
+      return x
+
+
+.. codediff::
+  :title_left: Haiku
+  :title_right: Flax
+  :sync:
+
+  def forward(x, training: bool):
+    return Model(256, 10)(x, training)
+
+  model = hk.transform_with_state(forward)
+
+  ---
+
+  ...
+
+
+  model = Model(256, 10)
+
+Initialization
+---------------
+
+.. codediff::
+  :title_left: Haiku
+  :title_right: Flax
+  :sync:
+
+  sample_x = jax.numpy.ones((1, 784))
+  params, state = model.init(
+      PRNGKey(0),
+      sample_x, training=False # <== inputs
+  )
+  ...
+
+  ---
+
+  sample_x = jax.numpy.ones((1, 784))
+  variables = model.init(
+      PRNGKey(0),
+      sample_x, training=True # <== inputs
+  )
+  params, batch_stats = variables["params"], variables["batch_stats"]
+
+.. codediff::
+  :title_left: Haiku
+  :title_right: Flax
+  :sync:
+
+  ...
+  {
+    'model/block/batch_norm': {
+      'offset': (1, 256),
+      'scale': (1, 256),
+    },
+    'model/block/linear': {
+      'b': (256,),
+      'w': (784, 256),
+    },
+    'model/linear': {
+      'b': (10,),
+      'w': (256, 10),
+    }
+  }
+  ...
+
+  ---
+
+  FrozenDict({
+    Block_0: {
+      BatchNorm_0: {
+        bias: (256,),
+        scale: (256,),
+      },
+      Dense_0: {
+        bias: (256,),
+        kernel: (784, 256),
+      },
+    },
+    Dense_0: {
+      bias: (10,),
+      kernel: (256, 10),
+    },
+  })
+
+Training
+---------
+
+.. codediff::
+  :title_left: Haiku
+  :title_right: Flax
+  :sync:
+
+  def train_step(key, params, state, inputs, labels):
+    def loss_fn(params):
+        logits, new_state = model.apply(
+            params, state,
+            key,
+            inputs, training=True # <== inputs
+            #
+        )
+        loss = optax.softmax_cross_entropy_with_integer_labels(logits, labels).mean()
+        return loss, new_state
+
+    grads, state = jax.grad(loss_fn, has_aux=True)(params)
+    params = jax.tree_map(lambda p, g: p - 0.1 * g, params, grads)
+
+    return params, state
+
+  ---
+
+  def train_step(key, params, batch_stats, inputs, labels):
+    def loss_fn(params):
+        logits, updates = model.apply(
+            {'params': params, 'batch_stats': batch_stats},
```

Comment:
Silly idea: if we're in a world where Transformers are all you need, and transformers don't use batch norm, would it make sense to have the simple example be simpler and just have `params` and no `batch_stats` and have a second subsection that can go in more depth about state and its myriad uses (including batch_norm) with this worked example?

Alternatively, can we simplify this code by passing in `variables` instead of `params` and `batch_stats` to `train_step` separately, and switching to `mutable='batch_stats'` to make this code have a lot less syntactical overhead?

Reply:
So this is what I had in mind: using `Dropout` and `BatchNorm` usually reveals most hidden knobs in most JAX frameworks, even if the model if very simple. So this example, while simple, shows 90% of what most users struggle with (managing state and random state).

I think this example is very worthwhile because even in transformers you will have the `cache` collection, or use the `intermediates` collection if the `sow` stuff, so dealing with the `batch_stats` collection serves as a good starting point even if the layer is not used in transformers.

That said, if we feel starting with something even simpler is better I can just remove the `BatchNorm` layer.

Reply:
in the transformer world you have to deal w. cache - I think in a lot of haiku codebases they just thread it through? (not sure I'd have to look) but they could in theory handle it with their state mechanism.

Reply:
Should we progressively increase the complexity of the examples in this guide? e.g. start with a regular model, then add Dropout, then add BatchNorm (or any other state). Might be friendlier but longer, this is more concise, straight to the crux of the matter.